File: src/main/java/com/tealcube/minecraft/bukkit/mythicdrops/tiers/MythicTier.java
package com.tealcube.minecraft.bukkit.mythicdrops.tiers;

/*
 * #%L
 * MythicDrops
 * %%
 * Copyright (C) 2013 - 2015 TealCube
 * %%
 * Permission to use, copy, modify, and/or distribute this software for any purpose with or without fee is hereby granted,
 * provided that the above copyright notice and this permission notice appear in all copies.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
 * INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF
 * THIS SOFTWARE.
 * #L%
 */


import com.tealcube.minecraft.bukkit.mythicdrops.api.enchantments.MythicEnchantment;
import com.tealcube.minecraft.bukkit.mythicdrops.api.tiers.Tier;

import org.bukkit.ChatColor;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

public final class MythicTier implements Tier {

    private final String name;
    private String displayName;
    private ChatColor displayColor;
    private ChatColor identificationColor;
    private List<String> baseLore;
    private List<String> bonusLore;
    private int minimumBonusLore;
    private int maximumBonusLore;
    private Set<MythicEnchantment> baseEnchantments;
    private Set<MythicEnchantment> bonusEnchantments;
    private boolean safeBaseEnchantments;
    private boolean safeBonusEnchantments;
    private boolean allowHighBaseEnchantments;
    private boolean allowHighBonusEnchantments;
    private int minimumBonusEnchantments;
    private int maximumBonusEnchantments;
    private double minimumDurabilityPercentage;
    private double maximumDurabilityPercentage;
    private List<String> allowedItemGroups;
    private List<String> disallowedItemGroups;
    private List<String> allowedItemIds;
    private List<String> disallowedItemIds;
    private int minimumSockets;
    private int maximumSockets;
    private double chanceToHaveSockets;
    private boolean broadcastOnFind;
    private double spawnChance;
    private double dropChance;
    private double identifyChance;
    private int optimalDistance = -1;
    private int maximumDistance = -1;
    private boolean infiniteDurability;

    protected MythicTier(String name) {
        this.name = name;
        baseLore = new ArrayList<>();
        bonusLore = new ArrayList<>();
        baseEnchantments = new HashSet<>();
        bonusEnchantments = new HashSet<>();
        allowedItemGroups = new ArrayList<>();
        disallowedItemGroups = new ArrayList<>();
        allowedItemIds = new ArrayList<>();
        disallowedItemIds = new ArrayList<>();
    }

    @Override
    public String getName() {
        return name;
    }

    @Override
    public String getDisplayName() {
        return displayName;
    }

    void setDisplayName(String displayName) {
        this.displayName = displayName;
    }

    @Override
    public ChatColor getDisplayColor() {
        return displayColor;
    }

    void setDisplayColor(ChatColor displayColor) {
        this.displayColor = displayColor;
    }

    @Override
    public ChatColor getIdentificationColor() {
        return identificationColor;
    }

    void setIdentificationColor(ChatColor identificationColor) {
        this.identificationColor = identificationColor;
    }

    @Override
    public List<String> getBaseLore() {
        return baseLore;
    }

    void setBaseLore(List<String> baseLore) {
        this.baseLore = baseLore;
    }

    @Override
    public List<String> getBonusLore() {
        return bonusLore;
    }

    void setBonusLore(List<String> bonusLore) {
        this.bonusLore = bonusLore;
    }

    @Override
    public int getMinimumBonusLore() {
        return minimumBonusLore;
    }

    void setMinimumBonusLore(int minimumBonusLore) {
        this.minimumBonusLore = minimumBonusLore;
    }

    @Override
    public int getMaximumBonusLore() {
        return maximumBonusLore;
    }

    void setMaximumBonusLore(int maximumBonusLore) {
        this.maximumBonusLore = maximumBonusLore;
    }

    @Override
    public Set<MythicEnchantment> getBaseEnchantments() {
        return baseEnchantments;
    }

    void setBaseEnchantments(Set<MythicEnchantment> baseEnchantments) {
        this.baseEnchantments = baseEnchantments;
    }

    @Override
    public Set<MythicEnchantment> getBonusEnchantments() {
        return bonusEnchantments;
    }

    void setBonusEnchantments(Set<MythicEnchantment> bonusEnchantments) {
        this.bonusEnchantments = bonusEnchantments;
    }

    @Override
    public boolean isSafeBaseEnchantments() {
        return safeBaseEnchantments;
    }

    void setSafeBaseEnchantments(boolean safeBaseEnchantments) {
        this.safeBaseEnchantments = safeBaseEnchantments;
    }

    @Override
    public boolean isSafeBonusEnchantments() {
        return safeBonusEnchantments;
    }

    void setSafeBonusEnchantments(boolean safeBonusEnchantments) {
        this.safeBonusEnchantments = safeBonusEnchantments;
    }

    @Override
    public boolean isAllowHighBaseEnchantments() {
        return allowHighBaseEnchantments;
    }

    void setAllowHighBaseEnchantments(boolean allowHighBaseEnchantments) {
        this.allowHighBaseEnchantments = allowHighBaseEnchantments;
    }

    @Override
    public boolean isAllowHighBonusEnchantments() {
        return allowHighBonusEnchantments;
    }

    void setAllowHighBonusEnchantments(boolean allowHighBonusEnchantments) {
        this.allowHighBonusEnchantments = allowHighBonusEnchantments;
    }

    @Override
    public int getMinimumBonusEnchantments() {
        return minimumBonusEnchantments;
    }

    void setMinimumBonusEnchantments(int minimumBonusEnchantments) {
        this.minimumBonusEnchantments = minimumBonusEnchantments;
    }

    @Override
    public int getMaximumBonusEnchantments() {
        return maximumBonusEnchantments;
    }

    void setMaximumBonusEnchantments(int maximumBonusEnchantments) {
        this.maximumBonusEnchantments = maximumBonusEnchantments;
    }

    @Override
    public double getMaximumDurabilityPercentage() {
        return maximumDurabilityPercentage;
    }

    void setMaximumDurabilityPercentage(double maximumDurabilityPercentage) {
        this.maximumDurabilityPercentage = maximumDurabilityPercentage;
    }

    @Override
    public double getMinimumDurabilityPercentage() {
        return minimumDurabilityPercentage;
    }

    void setMinimumDurabilityPercentage(double minimumDurabilityPercentage) {
        this.minimumDurabilityPercentage = minimumDurabilityPercentage;
    }

    @Override
    public Map<String, Double> getWorldDropChanceMap() {
        return new HashMap<>();
    }

    void setWorldDropChanceMap(Map<String, Double> worldDropChanceMap) {
        // do nothing
    }

    @Override
    public double getDropChance() {
        return dropChance;
    }

    public void setDropChance(double dropChance) {
        this.dropChance = dropChance;
    }

    @Override
    public Map<String, Double> getWorldSpawnChanceMap() {
        return new HashMap<>();
    }

    void setWorldSpawnChanceMap(Map<String, Double> worldSpawnChanceMap) {
        // do nothing
    }

    @Override
    public double getSpawnChance() {
        return spawnChance;
    }

    public void setSpawnChance(double spawnChance) {
        this.spawnChance = spawnChance;
    }

    @Override
    public Map<String, Double> getWorldIdentifyChanceMap() {
        return new HashMap<>();
    }

    public void setWorldIdentifyChanceMap(Map<String, Double> map) {
        // do nothing
    }

    @Override
    public double getIdentifyChance() {
        return identifyChance;
    }

    public void setIdentifyChance(double identifyChance) {
        this.identifyChance = identifyChance;
    }

    @Override
    public List<String> getAllowedItemGroups() {
        return allowedItemGroups;
    }

    void setAllowedItemGroups(List<String> allowedItemGroups) {
        this.allowedItemGroups = allowedItemGroups;
    }

    @Override
    public List<String> getDisallowedItemGroups() {
        return disallowedItemGroups;
    }

    void setDisallowedItemGroups(List<String> disallowedItemGroups) {
        this.disallowedItemGroups = disallowedItemGroups;
    }

    @Override
    public List<String> getAllowedItemIds() {
        return allowedItemIds;
    }

    void setAllowedItemIds(List<String> allowedItemIds) {
        this.allowedItemIds = allowedItemIds;
    }

    @Override
    public List<String> getDisallowedItemIds() {
        return disallowedItemIds;
    }

    void setDisallowedItemIds(List<String> disallowedItemIds) {
        this.disallowedItemIds = disallowedItemIds;
    }

    @Override
    public int getMinimumSockets() {
        return minimumSockets;
    }

    public void setMinimumSockets(int minimumSockets) {
        this.minimumSockets = minimumSockets;
    }

    @Override
    public int getMaximumSockets() {
        return maximumSockets;
    }

    public void setMaximumSockets(int maximumSockets) {
        this.maximumSockets = maximumSockets;
    }

    @Override
    public double getChanceToHaveSockets() {
        return chanceToHaveSockets;
    }

    public void setChanceToHaveSockets(double chanceToHaveSockets) {
        this.chanceToHaveSockets = chanceToHaveSockets;
    }

    @Override
    public boolean isBroadcastOnFind() {
        return broadcastOnFind;
    }

    public void setBroadcastOnFind(boolean broadcastOnFind) {
        this.broadcastOnFind = broadcastOnFind;
    }

    @Override
    public int hashCode() {
        int result = name != null ? name.hashCode() : 0;
        result = 31 * result + (displayColor != null ? displayColor.hashCode() : 0);
        result = 31 * result + (identificationColor != null ? identificationColor.hashCode() : 0);
        return result;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof MythicTier)) {
            return false;
        }

        MythicTier that = (MythicTier) o;

        return displayColor == that.displayColor && identificationColor == that.identificationColor
               && !(name !=
                    null ? !name.equals(that.name) : that.name != null);
    }

    @Override
    public String toString() {
        return "MythicTier{" +
               "name='" + name + '\'' +
               '}';
    }

    @Override
    public int compareTo(Tier o) {
        if (o == null || this.equals(o)) {
            return 0;
        }
        return Double.compare(this.getSpawnChance(), o.getSpawnChance());
    }

    @Override
    public int getOptimalDistance() {
        return optimalDistance;
    }

    public void setOptimalDistance(int optimalDistance) {
        this.optimalDistance = optimalDistance;
    }

    @Override
    public int getMaximumDistance() {
        return maximumDistance;
    }

    public void setMaximumDistance(int maximumDistance) {
        this.maximumDistance = maximumDistance;
    }

    @Override
    public boolean isInfiniteDurability() {
        return infiniteDurability;
    }

    void setInfiniteDurability(boolean infiniteDurability) {
        this.infiniteDurability = infiniteDurability;
    }

}

File: src/main/kotlin/com/tealcube/minecraft/bukkit/mythicdrops/tiers/MythicTier.kt
/*
 * This file is part of MythicDrops, licensed under the MIT License.
 *
 * Copyright (C) 2019 Richard Harrah
 *
 * Permission is hereby granted, free of charge,
 * to any person obtaining a copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software,
 * and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS
 * OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
 * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package com.tealcube.minecraft.bukkit.mythicdrops.tiers

import com.squareup.moshi.JsonClass
import com.tealcube.minecraft.bukkit.mythicdrops.DEFAULT_REPAIR_COST
import com.tealcube.minecraft.bukkit.mythicdrops.api.attributes.MythicAttribute
import com.tealcube.minecraft.bukkit.mythicdrops.api.enchantments.MythicEnchantment
import com.tealcube.minecraft.bukkit.mythicdrops.api.errors.LoadingErrorManager
import com.tealcube.minecraft.bukkit.mythicdrops.api.items.ItemGroup
import com.tealcube.minecraft.bukkit.mythicdrops.api.items.ItemGroupManager
import com.tealcube.minecraft.bukkit.mythicdrops.api.tiers.Tier
import com.tealcube.minecraft.bukkit.mythicdrops.api.tiers.TierAttributes
import com.tealcube.minecraft.bukkit.mythicdrops.api.tiers.TierEnchantments
import com.tealcube.minecraft.bukkit.mythicdrops.api.tiers.TierItemTypes
import com.tealcube.minecraft.bukkit.mythicdrops.getChatColor
import com.tealcube.minecraft.bukkit.mythicdrops.getNonNullString
import com.tealcube.minecraft.bukkit.mythicdrops.getOrCreateSection
import com.tealcube.minecraft.bukkit.mythicdrops.utils.MinecraftVersions
import io.pixeloutlaw.minecraft.spigot.bandsaw.JulLoggerFactory
import io.pixeloutlaw.minecraft.spigot.mythicdrops.enumValueOrNull
import org.bukkit.ChatColor
import org.bukkit.Material
import org.bukkit.configuration.ConfigurationSection
import org.bukkit.inventory.ItemFlag

@JsonClass(generateAdapter = true)
data class MythicTier(
    override val name: String = "",
    override val displayName: String = "",
    override val displayColor: ChatColor = ChatColor.RESET,
    override val identifierColor: ChatColor = ChatColor.RESET,
    override val baseLore: List<String> = emptyList(),
    override val bonusLore: List<String> = emptyList(),
    override val minimumBonusLore: Int = 0,
    override val maximumBonusLore: Int = 0,
    override val enchantments: TierEnchantments = MythicTierEnchantments(),
    override val minimumDurabilityPercentage: Double = 0.0,
    override val maximumDurabilityPercentage: Double = 0.0,
    override val chanceToDropOnMonsterDeath: Double = 0.0,
    override val itemTypes: TierItemTypes = MythicTierItemTypes(),
    override val minimumDistanceFromSpawn: Int = -1,
    override val maximumDistanceFromSpawn: Int = -1,
    override val isUnbreakable: Boolean = false,
    override val identityWeight: Double = 0.0,
    override val weight: Double = 0.0,
    override val chanceToHaveSockets: Double = 0.0,
    override val minimumSockets: Int = 0,
    override val maximumSockets: Int = 0,
    override val isBroadcastOnFind: Boolean = false,
    override val attributes: TierAttributes = MythicTierAttributes(),
    override val itemDisplayNameFormat: String? = null,
    override val tooltipFormat: List<String>? = null,
    override val itemFlags: Set<ItemFlag> = emptySet(),
    override val chanceToHaveSocketExtenderSlots: Double = 0.0,
    override val minimumSocketExtenderSlots: Int = 0,
    override val maximumSocketExtenderSlots: Int = 0,
    override val repairCost: Int = DEFAULT_REPAIR_COST
) : Tier {
    companion object {
        private val logger = JulLoggerFactory.getLogger(MythicTier::class)

        @JvmStatic
        fun fromConfigurationSection(
            configurationSection: ConfigurationSection,
            key: String,
            itemGroupManager: ItemGroupManager,
            loadingErrorManager: LoadingErrorManager
        ): MythicTier? {
            val displayColor = configurationSection.getChatColor("display-color")?.let {
                if (it == ChatColor.WHITE && !MinecraftVersions.isAtLeastMinecraft116) {
                    logger.info("WHITE doesn't work due to a bug in Spigot, so we're replacing it with RESET instead")
                    ChatColor.RESET
                } else {
                    it
                }
            }
            if (displayColor == null) {
                logger.fine("displayColor == null, key=$key")
                loadingErrorManager.add("Not loading tier $key as it has an invalid display color")
                return null
            }
            val identifierColor = configurationSection.getChatColor("identifier-color")
            if (identifierColor == null) {
                logger.fine("identifierColor == null, key=$key")
                loadingErrorManager.add("Not loading tier $key as it has an invalid identifier color")
                return null
            }
            val itemTypesSection = configurationSection.getOrCreateSection("item-types")
            val tierItemTypes = MythicTierItemTypes.fromConfigurationSection(itemTypesSection, itemGroupManager)
            val attributesSection = configurationSection.getOrCreateSection("attributes")
            val tierAttributes = MythicTierAttributes.fromConfigurationSection(attributesSection)
            val enchantmentsSection = configurationSection.getOrCreateSection("enchantments")
            val tierEnchantments = MythicTierEnchantments.fromConfigurationSection(enchantmentsSection)
            val itemDisplayNameFormat = configurationSection.getString("item-display-name-format")
            val tooltipFormat = if (configurationSection.isList("tooltip-format")) {
                configurationSection.getStringList("tooltip-format").filterNotNull().toList()
            } else {
                null
            }
            val itemFlags =
                configurationSection.getStringList("item-flags").mapNotNull {
                    enumValueOrNull<ItemFlag>(
                        it
                    )
                }.toSet()
            return MythicTier(
                name = key,
                displayName = configurationSection.getNonNullString("display-name", key),
                displayColor = displayColor,
                identifierColor = identifierColor,
                baseLore = configurationSection.getStringList("lore.base-lore"),
                bonusLore = configurationSection.getStringList("lore.bonus-lore"),
                minimumBonusLore = configurationSection.getInt("lore.minimum-bonus-lore"),
                maximumBonusLore = configurationSection.getInt("lore.maximum-bonus-lore"),
                minimumDurabilityPercentage = configurationSection.getDouble("minimum-durability"),
                maximumDurabilityPercentage = configurationSection.getDouble("maximum-durability"),
                chanceToDropOnMonsterDeath = configurationSection.getDouble("chance-to-drop-on-monster-death"),
                itemTypes = tierItemTypes,
                minimumDistanceFromSpawn = configurationSection.getInt("minimum-distance-from-spawn", -1),
                maximumDistanceFromSpawn = configurationSection.getInt("maximum-distance-from-spawn", -1),
                isUnbreakable = configurationSection.getBoolean("unbreakable"),
                weight = configurationSection.getDouble("weight"),
                identityWeight = configurationSection.getDouble("identity-weight"),
                chanceToHaveSockets = configurationSection.getDouble("chance-to-have-sockets"),
                minimumSockets = configurationSection.getInt("minimum-sockets"),
                maximumSockets = configurationSection.getInt("maximum-sockets"),
                isBroadcastOnFind = configurationSection.getBoolean("broadcast-on-find"),
                attributes = tierAttributes,
                enchantments = tierEnchantments,
                itemDisplayNameFormat = itemDisplayNameFormat,
                tooltipFormat = tooltipFormat,
                itemFlags = itemFlags,
                chanceToHaveSocketExtenderSlots = configurationSection.getDouble(
                    "chance-to-have-socket-extender-slots"
                ),
                minimumSocketExtenderSlots = configurationSection.getInt("minimum-socket-extender-slots"),
                maximumSocketExtenderSlots = configurationSection.getInt("maximum-socket-extender-slots"),
                repairCost = configurationSection.getInt("repair-cost", DEFAULT_REPAIR_COST)
            )
        }
    }

    override val baseAttributes: Set<MythicAttribute>
        get() = attributes.baseAttributes
    override val bonusAttributes: Set<MythicAttribute>
        get() = attributes.bonusAttributes
    override val minimumBonusAttributes: Int
        get() = attributes.minimumBonusAttributes
    override val maximumBonusAttributes: Int
        get() = attributes.maximumBonusAttributes
    override val baseEnchantments: Set<MythicEnchantment>
        get() = enchantments.baseEnchantments
    override val bonusEnchantments: Set<MythicEnchantment>
        get() = enchantments.bonusEnchantments
    override val minimumBonusEnchantments: Int
        get() = enchantments.minimumBonusEnchantments
    override val maximumBonusEnchantments: Int
        get() = enchantments.maximumBonusEnchantments
    override val isSafeBaseEnchantments: Boolean
        get() = enchantments.isSafeBaseEnchantments
    override val isSafeBonusEnchantments: Boolean
        get() = enchantments.isSafeBonusEnchantments
    override val isSafeRelationEnchantments: Boolean
        get() = enchantments.isSafeRelationEnchantments
    override val isAllowHighBaseEnchantments: Boolean
        get() = enchantments.isAllowHighBaseEnchantments
    override val isAllowHighBonusEnchantments: Boolean
        get() = enchantments.isAllowHighBonusEnchantments
    override val isAllowHighRelationEnchantments: Boolean
        get() = enchantments.isAllowHighRelationEnchantments
    override val allowedItemGroups: Set<ItemGroup>
        get() = itemTypes.allowedItemGroups
    override val disallowedItemGroups: Set<ItemGroup>
        get() = itemTypes.disallowedItemGroups
    override val allowedMaterialIds: Set<Material>
        get() = itemTypes.allowedMaterialIds
    override val disallowedMaterialIds: Set<Material>
        get() = itemTypes.disallowedMaterialIds

    override fun compareTo(other: Tier): Int {
        return if (this == other) {
            0
        } else {
            weight.compareTo(other.weight)
        }
    }
}
